src/github.com/mozilla-services/pushgo/simplepush/dynamo_store.go

```diff
@@ -0,0 +1,507 @@
+	CountQuery([]dynamodb.AttributeComparison) (int64, error)
+	PutItem(string, string, []dynamodb.Attribute) (bool, error)
+	UpdateAttributes(*dynamodb.Key, []dynamodb.Attribute) (bool, error)
+	DeleteItem(*dynamodb.Key) (bool, error)
+	Query([]dynamodb.AttributeComparison) ([]map[string]*dynamodb.Attribute, error)
+	DeleteAttributes(*dynamodb.Key, []dynamodb.Attribute) (bool, error)
+}
+
+type dynamoServer interface {
+	NewTable(string, dynamodb.PrimaryKey) *dynamodb.Table
+	CreateTable(dynamodb.TableDescriptionT) (string, error)
+	ListTables() ([]string, error)
+}
+
+type DynamoDBStore struct {
+	logger        *SimpleLogger
+	region        aws.Region
+	auth          aws.Auth
+	pk            dynamodb.PrimaryKey
+	server        dynamoServer
+	table         dynamoTable
+	tablename     string
+	readProv      int64
+	writeProv     int64
+	closeSignal   chan bool
+	maxChannels   int
+	statusTimeout time.Duration
+	statusIdle    time.Duration
+}
+
+const (
+	RFC_ISO_8601  = "20060102T150405Z"
+	DYNAMO_PREFIX = "DynamoDB_20120810"
+)
+
+func (s *DynamoDBStore) waitUntilStatus(table dynamoTable, status string) (err error) {
+	done := make(chan bool)
+	timeout := time.After(s.statusTimeout)
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				desc, err := table.DescribeTable()
+				if err != nil {
+					return
+				}
+				if desc.TableStatus == status {
+					done <- true
+					return
+				}
+				time.Sleep(s.statusIdle)
+			}
+		}
+	}()
+	select {
+	case <-done:
+		err = nil
+		break
+	case <-timeout:
+		err = ErrDynamoDBTimeout
+		close(done)
+	}
+	return
+}
+
+func (s *DynamoDBStore) createTable() (err error) {
+	/* There are three primary fields for each row,
+	   uaid = UserAgentID, all registered items must have one of these
+	   chid = Either the ChannelID or " ". In this case " " fills the
+	   		"main" UAID entry, which can carry info like the Proprietary
+	   		Ping data, etc. CHID rows can be accessed by specifing a
+	   		QUERY with chid > " ".
+	   created = UTC().Unix() of the record creation. Used for cleanup.
+	*/
+	tableDesc := dynamodb.TableDescriptionT{
+		TableName: s.tablename,
+		AttributeDefinitions: []dynamodb.AttributeDefinitionT{
+			dynamodb.AttributeDefinitionT{UAID_LABEL, "S"},
+			dynamodb.AttributeDefinitionT{CHID_LABEL, "S"},
+			dynamodb.AttributeDefinitionT{MODD_LABEL, "N"},
+		},
+		KeySchema: []dynamodb.KeySchemaT{
+			dynamodb.KeySchemaT{UAID_LABEL, "HASH"},
+			dynamodb.KeySchemaT{CHID_LABEL, "RANGE"},
+		},
+		GlobalSecondaryIndexes: []dynamodb.GlobalSecondaryIndexT{
+			dynamodb.GlobalSecondaryIndexT{
+				IndexName: "uaid-modified-index",
+				KeySchema: []dynamodb.KeySchemaT{
+					dynamodb.KeySchemaT{UAID_LABEL, "HASH"},
+					dynamodb.KeySchemaT{MODD_LABEL, "RANGE"},
+				},
+				Projection: dynamodb.ProjectionT{"ALL"},
+				ProvisionedThroughput: dynamodb.ProvisionedThroughputT{
+					ReadCapacityUnits:  s.readProv,
+					WriteCapacityUnits: s.writeProv,
+				},
```

Comment:
yes, the secondary index is more for garbage collection than active use, so it can either be provisioned lower or simply ignored and explicit GC can occur during off-peak periods.